synedrion/src/tools/sss.rs

```diff
@@ -83,31 +99,42 @@ impl Polynomial {
 }
 
 #[derive(Debug, Clone, Serialize, Deserialize)]
-pub(crate) struct PublicPolynomial(Vec<Point>);
+#[serde(bound(deserialize = "for<'x> P: Deserialize<'x>"))]
+pub(crate) struct PublicPolynomial<P: SchemeParams>(Vec<Point<P>>);
 
-impl PublicPolynomial {
-    pub fn evaluate(&self, x: &ShareId) -> Point {
+impl<P> PublicPolynomial<P>
+where
+    P: SchemeParams,
+{
+    pub fn evaluate(&self, x: &ShareId<P>) -> Point<P> {
         evaluate_polynomial(&self.0, &x.0)
     }
 
-    pub fn coeff0(&self) -> Result<&Point, LocalError> {
+    pub fn coeff0(&self) -> Result<&Point<P>, LocalError> {
         self.0
             .first()
             .ok_or_else(|| LocalError::new("Invalid PublicPolynomial"))
     }
 }
 
-pub(crate) fn shamir_split(
+pub(crate) fn shamir_split<P>(
     rng: &mut impl CryptoRngCore,
-    secret: Secret<Scalar>,
+    secret: Secret<Scalar<P>>,
     threshold: usize,
-    indices: &[ShareId],
-) -> BTreeMap<ShareId, Secret<Scalar>> {
+    indices: &[ShareId<P>],
+) -> BTreeMap<ShareId<P>, Secret<Scalar<P>>>
+where
+    P: SchemeParams,
+{
     let polynomial = Polynomial::random(rng, secret, threshold);
     indices.iter().map(|idx| (*idx, polynomial.evaluate(idx))).collect()
 }
 
-pub(crate) fn interpolation_coeff(share_ids: &BTreeSet<ShareId>, share_id: &ShareId) -> Scalar {
+pub(crate) fn interpolation_coeff<P>(share_ids: &BTreeSet<ShareId<P>>, share_id: &ShareId<P>) -> Scalar<P>
+where
+    P: SchemeParams,
+    <P::Curve as CurveArithmetic>::Scalar: Copy,
```

Comment:
Do we really need this bound? Can we just clone?

Reply:
We don't, good catch!